### PR TITLE
Numtraits for rational

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

This PR implements...
- [x] is_zero for Q, MatQ, PolyOverQ
- [x] is_one for Q, PolyOverQ
- [x] is_square and is_identity for MatQ

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/rational/poly_over_q/properties.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+}
+
+#[cfg(test)]
+mod test_is_zero {
+    use super::PolyOverQ;
+    use std::str::FromStr;
+
+    /// ensure that is_zero returns `true` for the zero polynomial
+
```
Comment (MoogSven) [inline src/rational/poly_over_q/properties.rs:77]:
```suggestion
```

**Review by MoogSven — CHANGES_REQUESTED** [review]
Did not finish the review, since many of the suggestions are already part of #270.
Please check which of the suggestions from that pull request also apply to this one.

So far:
- [x] remove brackets at assert!(!([...])).
- [x] avoid unnecessarily complicated numbers
- [ ] Maybe use MatQ::new([...]) or similar functions instead of from_str if the values are not important.
- [ ] How to handle non square matrices

File: src/rational/mat_q/properties.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        let ident = MatQ::from_str("[[1, 0],[0, 1]]").unwrap();
+
+        assert!(ident.is_identity());
+    }
+
+    /// ensure that is_identity returns `false` for non-identity matrices
+    #[test]
+    fn identity_rejection() {
+        let small = MatQ::from_str("[[0, 0],[2/81, 0]]").unwrap();
+        let large = MatQ::from_str(&format!("[[1, 0],[0, {}]]", (u128::MAX - 1) / 2 + 2)).unwrap();
```
[the remark this answers is not in the document]
Comment (3pmile) [reply src/rational/mat_q/properties.rs:78]: this is 2^127 which is 0 in the for the last 64 bit

File: src/rational/mat_q/properties.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        let zero = MatQ::from_str("[[0, 0],[0, 0]]").unwrap();
+
+        assert!(zero.is_zero());
+    }
+
+    /// ensure that is_zero returns `false` for non-zero matrices
+    #[test]
+    fn zero_rejection() {
+        let small = MatQ::from_str("[[0, 7/8],[2, 0]]").unwrap();
+        let large = MatQ::from_str(&format!("[[0, 0],[{}, 0]]", (u128::MAX - 1) / 2 + 1)).unwrap();
```
[the remark this answers is not in the document]
Comment (3pmile) [reply src/rational/mat_q/properties.rs:105]: same reason as above, the last bits are all 0

File: src/rational/mat_q/properties.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    ///
+    /// ```
+    /// use qfall_math::rational::MatQ;
+    /// use std::str::FromStr;
+    ///
+    /// let value = MatQ::from_str("[[1, 0],[0, 1]]").unwrap();
+    /// assert!(value.is_identity())
+    /// ```
+    pub fn is_identity(&self) -> bool {
+        self.is_square() && 1 == unsafe { fmpq_mat_is_one(&self.matrix) }
```
[the remark this answers is not in the document]
Comment (3pmile) [reply src/rational/mat_q/properties.rs:27]: This implementation is equal to other is_identity implementations

**Pull request closed without merging**